torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -1397,6 +1397,62 @@ def test_single_threaded_rref_owner(self):
         dist.barrier()
         rpc.shutdown()
 
+    @requires_process_group_agent("PROCESS_GROUP rpc backend specific test, skip")
+    def test_single_threaded_rref_to_here(self):
+        # This test aims to verify if the server can handle all internal RPC
+        # messages using just one thread.
+        caller_rank = 0
+        callee_rank = 1
+        rpc_backend_options = rpc.ProcessGroupRpcBackendOptions(
+            init_method=self.rpc_backend_options.init_method,
+            num_send_recv_threads=1
+        ) if self.rank == callee_rank else self.rpc_backend_options
+
+        rpc.init_rpc(
+            name=worker_name(self.rank),
+            backend=self.rpc_backend,
+            rank=self.rank,
+            world_size=self.world_size,
+            rpc_backend_options=rpc_backend_options,
+        )
+
+        if self.rank == caller_rank:
+            dst = worker_name(callee_rank)
+            rrefs = []
+
+            # makes sure there is no existing OwnerRRefs on dst
+            info = rpc.rpc_sync(dst, get_rref_debug_info)
+            self.assertEqual(0, int(info["num_owner_rrefs"]))
```

Comment:
test looks fine to me, although, is there a way to check the behavior by ensuring that the remote end processes `rref.to_here()` before `rpc.remote()` 

I'm assuming the reason for this PR is that before with single threaded, the thread could block on a `to_here()`. But then it would never process the `rpc.remote()`, hence hanging forever.

Reply:
> I'm assuming the reason for this PR is that before with single threaded, the thread could block on a to_here(). But then it would never process the rpc.remote(), hence hanging forever.

Yes. `to_here` should not block, as it does not contain any user code.

>  is there a way to check the behavior by ensuring that the remote end processes rref.to_here() before rpc.remote()

Didn't figure out a way to test this with the current code. We need to somehow delay sending the first `rpc.remote` message and send the `to_here` message first. We might need a dedicated test RPC agent which allows application to control its send queue.